java-checks/src/main/java/org/sonar/java/checks/SillyEqualsCheck.java

```diff
@@ -0,0 +1,120 @@
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02
+ */
+package org.sonar.java.checks;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.Iterables;
+import org.sonar.api.server.rule.RulesDefinition;
+import org.sonar.check.Priority;
+import org.sonar.check.Rule;
+import org.sonar.java.checks.methods.AbstractMethodDetection;
+import org.sonar.java.checks.methods.MethodInvocationMatcher;
+import org.sonar.plugins.java.api.semantic.Type;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MemberSelectExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.ActivatedByDefault;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.List;
+
+@Rule(
+  key = "S2159",
+  name = "Silly equality checks should not be made",
+  tags = {"bug", "unused"},
+  priority = Priority.MAJOR)
+@ActivatedByDefault
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.LOGIC_RELIABILITY)
+@SqaleConstantRemediation("15min")
+public class SillyEqualsCheck extends AbstractMethodDetection {
+
+  private static final String JAVA_LANG_OBJECT = "java.lang.Object";
+
+  private static final String MESSAGE = "Remove this call to \"equals\"; comparisons between unrelated types always return false.";
+
+  @Override
+  protected List<MethodInvocationMatcher> getMethodInvocationMatchers() {
+    return ImmutableList.of(MethodInvocationMatcher.create()
+      .name("equals")
+      .addParameter(JAVA_LANG_OBJECT));
+  }
+
+  @Override
+  protected void onMethodFound(MethodInvocationTree tree) {
+    ExpressionTree firstArgument = Iterables.getOnlyElement(tree.arguments());
+    Type argumentType = firstArgument.symbolType().erasure();
+    Type ownerType = getMethodOwnerType(tree).erasure();
+    if (isLiteralNull(firstArgument)) {
+      addIssue(tree, "Remove this call to \"equals\"; comparisons against null always return false; consider using '== null' to check for nullity.");
+    } else if (ownerType.isArray()) {
+      checkWhenOwnerIsArray(tree, (Type.ArrayType) ownerType, argumentType);
+    } else {
+      checkWhenOwnerIsNotArray(tree, ownerType, argumentType);
+    }
+  }
+
+  private void checkWhenOwnerIsArray(Tree tree, Type.ArrayType ownerType, Type argumentType) {
+    if (argumentType.isArray()) {
+      if (areNotRelated(((Type.ArrayType) ownerType).elementType(), ((Type.ArrayType) argumentType).elementType())) {
```

Comment:
:new: ![MINOR](https://raw.githubusercontent.com/henryju/image-hosting/master/minor.png)Remove this unnecessary cast to "ArrayType".[![...](https://raw.githubusercontent.com/henryju/image-hosting/master/more.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS1905)